File: scripts/python/src/tools/transcript/menu.py
import collections
import contextlib
import shutil
import sys

from tools.transcript.screen import (
    ABORT,
    BLANK,
    BOLD,
    CLEAR,
    CURSOR,
    CYAN,
    DIM,
    ERASE,
    GAP,
    HIDE,
    INDENT,
    SHOW,
    colors_for,
    compose,
    flush,
    normalise,
    scripted,
    tty_keys,
)

GUTTER = "    "
DIGITS = "123456789"

HEADER = 4
SPARE = 1
MIN_ROWS = 3
MIN_WIDTH = 24

HINT = "↑/↓ move | ↩ select | q quit"
DEEP_HINT = "↑/↓ move | ←/→ level | ↩ select | q quit"

Pick = collections.namedtuple("Pick", "kind index option")

_stream = None
_height = 0


class Column:
    def __init__(self, options, details=None, default=0, kind=""):
        self.options = list(options)
        details = list(details or [])[: len(self.options)]
        self.details = details + [""] * (len(self.options) - len(details))
        self.kind = kind
        self.index = min(max(default, 0), max(len(self.options) - 1, 0))

    def label_width(self):
        return max(len(option) for option in self.options)

    def width(self, detailed):
        room = len(CURSOR) + self.label_width()
        if detailed and any(self.details):
            room += len(GAP) + max(len(detail) for detail in self.details)
        return room


class Cascade:
    def __init__(self, title, expand, out, color_on):
        self.title = title
        self.expand = expand
        self.out = out
        self.color_on = color_on
        self.columns = []
        self.opened = {}
        self.height = 0

    def picks(self, depth=None):
        columns = self.columns if depth is None else self.columns[:depth]
        return tuple(Pick(one.kind, one.index, one.options[one.index]) for one in columns)

    def open(self, path):
        if path not in self.opened:
            self.opened[path] = self.expand(path)
        return self.opened[path]

    def descend(self):
        child = self.open(self.picks())
        if child is None or not child.options:
            return False
        self.columns.append(child)
        return True

    def restore(self, start):
        for depth, index in enumerate(start):
            column = self.columns[depth]
            column.index = min(max(index, 0), len(column.options) - 1)
            if depth + 1 >= len(start) or not self.descend():
                return

    def stack(self, limit):
        offsets = []
        for position, column in enumerate(self.columns):
            if not position:
                offsets.append(0)
                continue
            previous = self.columns[position - 1]
            offset = min(offsets[-1] + previous.index, limit - len(column.options))
            offsets.append(max(0, offset))
        return offsets, max(o + len(c.options) for o, c in zip(offsets, self.columns))

    def cell(self, column, index, label, active):
        if index < 0 or index >= len(column.options):
            return [], 0
        chosen = index == column.index
        if active:
            mark = tint = CYAN + BOLD if chosen else ""
        else:
            mark, tint = (CYAN if chosen else ""), ("" if chosen else DIM)
        option = column.options[index]
        segments = [(CURSOR if chosen else BLANK, mark), (option, tint)]
        used = len(CURSOR) + len(option)
        detail = column.details[index] if active else ""
        if detail:
            segments.append((" " * (label - len(option)) + GAP, ""))
            segments.append((detail, DIM))
            used += label - len(option) + len(GAP) + len(detail)
        return segments, used

    def row(self, index, offsets, widths, first, avail):
        last = len(self.columns) - 1
        segments = [(INDENT, "")]
        for position in range(first, last + 1):
            if position > first:
                segments.append((GUTTER, ""))
            column = self.columns[position]
            active = position == last
            cell, used = self.cell(column, index - offsets[position], column.label_width(), active)
            segments.extend(cell)
            if not active:
                segments.append((" " * (widths[position] - used), ""))
        return compose(segments, avail, self.color_on)[0].rstrip()

    def trimmed(self, widths, avail):
        first = 0
        while first < len(self.columns) - 1:
            gutters = len(GUTTER) * (len(self.columns) - 1 - first)
            if len(INDENT) + sum(widths[first:]) + gutters <= avail:
                break
            first += 1
        return first

    def frame(self):
        avail = max(shutil.get_terminal_size().columns - 1, MIN_WIDTH)
        budget = max(shutil.get_terminal_size().lines - HEADER - SPARE, MIN_ROWS)
        offsets, height = self.stack(budget)
        last = len(self.columns) - 1
        widths = [column.width(position == last) for position, column in enumerate(self.columns)]
        first = self.trimmed(widths, avail)
        if first:
            lift = min(offsets[first:])
            offsets = [offset - lift for offset in offsets]
            height = max(offsets[at] + len(self.columns[at].options) for at in range(first, last + 1))
        heading = [(INDENT, ""), (self.title, BOLD)]
        if first:
            walked = " › ".join(one.options[one.index] for one in self.columns[:first])
            heading.append((f"  ‹ {walked}", DIM))
        hint = DEEP_HINT if len(self.columns) > 1 else HINT
        lines = [
            "",
            compose(heading, avail, self.color_on)[0],
            compose([(INDENT, ""), (hint, DIM)], avail, self.color_on)[0],
            "",
        ]
        lines += [self.row(index, offsets, widths, first, avail) for index in range(height)]
        return lines

    def draw(self):
        lines = self.frame()
        write = self.out.write
        if self.height:
            write(f"\033[{self.height}A")
        for line in lines:
            write(line + CLEAR + "\n")
        if len(lines) < self.height:
            write(ERASE)
        self.height = len(lines)
        flush(self.out)

    def wipe(self):
        if not self.height:
            return
        self.out.write(f"\033[{self.height}A" + ERASE)
        self.height = 0
        flush(self.out)

    def summary(self, picks):
        avail = max(shutil.get_terminal_size().columns - 1, MIN_WIDTH)
        steps = [pick.option for pick in picks]
        segments = [(INDENT, ""), (self.title, DIM), (" — ", DIM)]
        if len(steps) > 1:
            segments.append((" › ".join(steps[:-1]) + " › ", DIM))
        segments.append((steps[-1], CYAN))
        return compose(segments, avail, self.color_on)[0]

    def run(self, reader, start):
        root = self.open(())
        if root is None or not root.options:
            return None
        self.columns.append(root)
        self.restore(tuple(start))
        while True:
            self.draw()
            key = normalise(reader())
            if key in ABORT:
                return None
            column = self.columns[-1]
            count = len(column.options)
            if key == "up":
                column.index = (column.index - 1) % count
            elif key == "down":
                column.index = (column.index + 1) % count
            elif key == "left":
                if len(self.columns) > 1:
                    self.columns.pop()
            elif key in ("right", "enter"):
                if not self.descend() and key == "enter":
                    return list(self.picks())
            elif key and key in DIGITS and int(key) <= count:
                column.index = int(key) - 1


def cascade(title, expand, start=(), keys=None, out=None):
    global _stream, _height
    stream = sys.stdout if out is None else out
    if out is None and not stream.isatty():
        return None
    runner = Cascade(title, expand, stream, colors_for(stream))
    with contextlib.ExitStack() as stack:
        if keys is None:
            try:
                reader = stack.enter_context(tty_keys())
            except OSError:
                return None
        else:
            reader = scripted(keys)
        stack.callback(flush, stream)
        stack.callback(stream.write, SHOW)
        stream.write(HIDE)
        try:
            picks = runner.run(reader, start)
        except KeyboardInterrupt:
            picks = None
        finally:
            runner.wipe()
        _stream, _height = stream, 0
        if picks:
            stream.write(runner.summary(picks) + CLEAR + "\n")
            _height = 1
        return picks


def erase():
    global _height
    if not _height or _stream is None:
        return
    _stream.write(f"\033[{_height}A" + ERASE)
    flush(_stream)
    _height = 0

File: scripts/rust/crates/sysinfo/src/bench/menu.rs
use super::{
    record::{ANY, CLEAN, Run},
    report,
    store::Store,
};
use std::collections::BTreeSet;
use workstation::{Key, Screen};

#[derive(Clone, Debug)]
pub struct Pick {
    pub kind: String,
    pub option: String,
}
#[derive(Clone, Debug)]
pub struct Column {
    pub kind: String,
    pub title: String,
    pub options: Vec<(String, String)>,
    pub index: usize,
}
impl Column {
    fn new(kind: &str, title: &str, options: Vec<(String, String)>) -> Self {
        Self {
            kind: kind.into(),
            title: title.into(),
            options,
            index: 0,
        }
    }
    fn picked(&self) -> Pick {
        Pick {
            kind: self.kind.clone(),
            option: self.options[self.index].0.clone(),
        }
    }
}
fn selected<'a>(picks: &'a [Pick], kind: &str) -> &'a str {
    picks
        .iter()
        .find(|pick| pick.kind == kind)
        .map(|pick| pick.option.as_str())
        .unwrap_or("")
}
fn note(message: &str) -> Column {
    Column::new("note", "", vec![(message.into(), String::new())])
}
fn simple(kind: &str, title: &str, options: Vec<String>) -> Column {
    Column::new(
        kind,
        title,
        options
            .into_iter()
            .map(|option| (option, String::new()))
            .collect(),
    )
}
pub fn choose(title: &str, options: &[String]) -> Result<Option<usize>, String> {
    let found = cascade(title, |picks| {
        if picks.is_empty() {
            Some(simple("choice", title, options.to_vec()))
        } else {
            None
        }
    })?;
    Ok(found.and_then(|picks| {
        picks
            .last()
            .and_then(|pick| options.iter().position(|option| option == &pick.option))
    }))
}
pub fn cascade(
    title: &str,
    expand: impl Fn(&[Pick]) -> Option<Column>,
) -> Result<Option<Vec<Pick>>, String> {
    super::require_terminal(title)?;
    let Some(root) = expand(&[]).filter(|column| !column.options.is_empty()) else {
        return Ok(None);
    };
    let Some(mut screen) = Screen::open().map_err(|e| e.to_string())? else {
        return Err("terminal unavailable".into());
    };
    let mut columns = vec![root];
    loop {
        let size = screen.size().unwrap_or((80, 24));
        screen
            .draw(&frame(title, &columns, size.0, size.1))
            .map_err(|e| e.to_string())?;
        let key = screen.key().map_err(|e| e.to_string())?;
        match key {
            Key::Escape | Key::Interrupt | Key::Char('q') => {
                screen.clear().map_err(|e| e.to_string())?;
                return Ok(None);
            }
            Key::Left | Key::Backspace | Key::Char('h') => {
                if columns.len() > 1 {
                    columns.pop();
                }
            }
            Key::Up | Key::Char('k') => {
                let column = columns.last_mut().unwrap();
                column.index = (column.index + column.options.len() - 1) % column.options.len();
            }
            Key::Down | Key::Char('j') => {
                let column = columns.last_mut().unwrap();
                column.index = (column.index + 1) % column.options.len();
            }
            Key::Home => columns.last_mut().unwrap().index = 0,
            Key::End => {
                let column = columns.last_mut().unwrap();
                column.index = column.options.len() - 1;
            }
            Key::PageUp | Key::PageDown => {
                let column = columns.last_mut().unwrap();
                let step = size.1.saturating_sub(6).max(1);
                column.index = if key == Key::PageUp {
                    column.index.saturating_sub(step)
                } else {
                    (column.index + step).min(column.options.len() - 1)
                };
            }
            Key::Enter | Key::Right | Key::Char('l') => {
                let picks = columns.iter().map(Column::picked).collect::<Vec<_>>();
                if let Some(child) = expand(&picks).filter(|column| !column.options.is_empty()) {
                    columns.push(child);
                } else if key == Key::Enter {
                    screen.clear().map_err(|e| e.to_string())?;
                    drop(screen);
                    println!(
                        "  {title} — {}",
                        picks
                            .iter()
                            .map(|pick| pick.option.as_str())
                            .collect::<Vec<_>>()
                            .join(" › ")
                    );
                    return Ok(Some(picks));
                }
            }
            Key::Char(digit) if digit.is_ascii_digit() && digit != '0' => {
                let index = (digit as u8 - b'1') as usize;
                let column = columns.last_mut().unwrap();
                if index < column.options.len() {
                    column.index = index;
                }
            }
            _ => {}
        }
    }
}
pub fn frame(title: &str, columns: &[Column], width: usize, height: usize) -> Vec<String> {
    let width = width.saturating_sub(1).max(1);
    let room = height.saturating_sub(6).max(1);
    let active = columns.len() - 1;
    let widths = columns
        .iter()
        .enumerate()
        .map(|(index, column)| {
            column
                .options
                .iter()
                .map(|(name, detail)| {
                    unicode_width::UnicodeWidthStr::width(name.as_str())
                        + 2
                        + if index == active && !detail.is_empty() {
                            2 + unicode_width::UnicodeWidthStr::width(detail.as_str())
                        } else {
                            0
                        }
                })
                .max()
                .unwrap_or(0)
                .max(unicode_width::UnicodeWidthStr::width(column.title.as_str()))
        })
        .collect::<Vec<_>>();
    let mut first = 0;
    while first < active && 2 + widths[first..].iter().sum::<usize>() + (active - first) * 3 > width
    {
        first += 1;
    }
    let heading = if first > 0 {
        format!(
            "  {title}  ‹ {}",
            columns[..first]
                .iter()
                .map(|column| column.options[column.index].0.as_str())
                .collect::<Vec<_>>()
                .join(" › ")
        )
    } else {
        format!("  {title}")
    };
    let mut lines = vec![
        String::new(),
        fit(&heading, width),
        fit("  ↑/↓ move | ←/→ level | ↩ select | q quit", width),
        String::new(),
    ];
    let mut headers = String::from("  ");
    for (index, column) in columns.iter().enumerate().skip(first) {
        if index > first {
            headers.push_str("   ");
        }
        headers.push_str(&column.title);
        if index < active {
            headers.push_str(
                &" ".repeat(
                    widths[index].saturating_sub(unicode_width::UnicodeWidthStr::width(
                        column.title.as_str(),
                    )),
                ),
            );
        }
    }
    lines.push(fit(&headers, width));
    let shown = columns[first..]
        .iter()
        .map(|column| column.options.len().min(room))
        .max()
        .unwrap_or(0);
    for row in 0..shown {
        let mut line = String::from("  ");
        for (index, column) in columns.iter().enumerate().skip(first) {
            if index > first {
                line.push_str("   ");
            }
            let start = column.index.saturating_sub(room.saturating_sub(1));
            let item = start + row;
            let cell = if let Some((option, detail)) = column.options.get(item) {
                format!(
                    "{}{}{}",
                    if item == column.index { "❯ " } else { "  " },
                    option,
                    if index == active && !detail.is_empty() {
                        format!("  {detail}")
                    } else {
                        String::new()
                    }
                )
            } else {
                String::new()
            };
            line.push_str(&cell);
            if index < active {
                line.push_str(
                    &" ".repeat(
                        widths[index]
                            .saturating_sub(unicode_width::UnicodeWidthStr::width(cell.as_str())),
                    ),
                );
            }
        }
        lines.push(fit(&line, width));
    }
    lines
}
fn fit(text: &str, width: usize) -> String {
    let mut result = String::new();
    let mut used = 0;
    for ch in text.chars() {
        let size = unicode_width::UnicodeWidthChar::width(ch).unwrap_or(0);
        if used + size > width {
            break;
        }
        result.push(ch);
        used += size;
    }
    result
}
const MENU: [(&str, &str); 8] = [
    ("run", "measure this machine now"),
    ("show", "inspect a stored run"),
    ("health", "warnings derived from benchmark history"),
    ("list", "stored runs"),
    ("compare", "two runs side by side"),
    ("trend", "one metric over time"),
    ("baseline", "set or clear the reference run"),
    ("prune", "thin old runs"),
];
const MODES: [(&str, &str); 4] = [
    (
        "machine vs machine",
        "the same metric on two different machines",
    ),
    (
        "before vs after upgrade",
        "two hardware configurations of one machine",
    ),
    ("distro vs distro", "two installations on one machine"),
    ("pick two runs", "choose both sides by hand"),
];
struct History {
    runs: Vec<Run>,
    hosts: Vec<String>,
}
impl History {
    fn host<'a>(&'a self, picks: &'a [Pick]) -> &'a str {
        let host = selected(picks, "host");
        if host.is_empty() {
            self.hosts.first().map(String::as_str).unwrap_or("")
        } else {
            host
        }
    }
    fn run_column(&self, host: &str, clean: bool, title: &str, kind: &str) -> Column {
        let found = self
            .runs
            .iter()
            .filter(|run| (host.is_empty() || run.host == host) && (!clean || run.grade == "clean"))
            .map(|run| {
                (
                    format!("{}:{}", run.host, run.run_id),
                    report::describe_run(run),
                )
            })
            .collect::<Vec<_>>();
        if found.is_empty() {
            note("no runs recorded")
        } else {
            Column::new(kind, title, found)
        }
    }
    fn grouping(picks: &[Pick]) -> (&'static str, &'static str) {
        if selected(picks, "compare") == "before vs after upgrade" {
            ("epoch", "hardware configuration")
        } else {
            ("install", "installation")
        }
    }
    fn group_keys(&self, host: &str, kind: &str) -> Vec<String> {
        let mut keys = Vec::new();
        for run in self.runs.iter().filter(|run| run.host == host) {
            let key = if kind == "epoch" {
                run.epoch()
            } else {
                run.os_id().into()
            };
            if !keys.contains(&key) {
                keys.push(key);
            }
        }
        keys
    }
    fn pair(&self, picks: &[Pick], side: &str) -> Column {
        let host = self.host(picks);
        let (kind, noun) = Self::grouping(picks);
        let keys = self.group_keys(host, kind);
        if keys.len() < 2 {
            return note(&format!("{host} has only one {noun} on record"));
        }
        let selected_a = selected(picks, &format!("{kind}-a"));
        simple(
            &format!("{kind}-{side}"),
            match (kind, side) {
                ("epoch", "a") => "earlier configuration",
                ("epoch", _) => "later configuration",
                (_, "a") => "first installation",
                _ => "second installation",
            },
            keys.into_iter()
                .filter(|key| side == "a" || key != selected_a)
                .collect(),
        )
    }
    fn after_host(&self, flow: &str, picks: &[Pick]) -> Column {
        let host = self.host(picks);
        match flow {
            "trend" => {
                let keys = self
                    .runs
                    .iter()
                    .filter(|run| run.host == host && run.grade == "clean")
                    .flat_map(|run| run.metrics.iter().map(|metric| metric.key.clone()))
                    .collect::<BTreeSet<_>>();
                if keys.is_empty() {
                    note(&format!("{host} has no clean runs"))
                } else {
                    simple("metric", "which metric?", keys.into_iter().collect())
                }
            }
            "baseline" => self.run_column(host, true, "use which run as the baseline?", "run"),
            _ => self.pair(picks, "a"),
        }
    }
    fn opening(&self, flow: &str, picks: &[Pick]) -> Option<Column> {
        match flow {
            "show" => Some(self.run_column("", false, "show which run?", "run")),
            "compare" => Some(Column::new(
                "compare",
                "compare what?",
                MODES
                    .iter()
                    .map(|(a, b)| (a.to_string(), b.to_string()))
                    .collect(),
            )),
            "trend" | "baseline" => Some(if self.hosts.len() > 1 {
                simple("host", "which machine?", self.hosts.clone())
            } else {
                self.after_host(flow, picks)
            }),
            _ => None,
        }
    }
    fn expand(&self, flow: &str, picks: &[Pick]) -> Option<Column> {
        if picks.is_empty() {
            return if flow.is_empty() {
                Some(Column::new(
                    "menu",
                    "",
                    MENU.iter()
                        .map(|(a, b)| (a.to_string(), b.to_string()))
                        .collect(),
                ))
            } else {
                self.opening(flow, picks)
            };
        }
        let last = picks.last()?;
        match last.kind.as_str() {
            "menu" => self.opening(&last.option, picks),
            "compare" => Some(match last.option.as_str() {
                "machine vs machine" => {
                    if self.hosts.len() < 2 {
                        note("two machines are needed; only one has runs")
                    } else {
                        simple("host-a", "first machine", self.hosts.clone())
                    }
                }
                "pick two runs" => self.run_column("", false, "left run", "run-a"),
                _ => {
                    if self.hosts.len() > 1 {
                        simple("host", "which machine?", self.hosts.clone())
                    } else {
                        self.pair(picks, "a")
                    }
                }
            }),
            "host" => Some(self.after_host(
                if flow.is_empty() {
                    selected(picks, "menu")
                } else {
                    flow
                },
                picks,
            )),
            "host-a" => Some(simple(
                "host-b",
                "second machine",
                self.hosts
                    .iter()
                    .filter(|name| *name != &last.option)
                    .cloned()
                    .collect(),
            )),
            "run-a" => Some(self.run_column("", false, "right run", "run-b")),
            "epoch-a" | "install-a" => Some(self.pair(picks, "b")),
            _ => None,
        }
    }
    fn sides(&self, store: &Store, picks: &[Pick]) -> Result<(Run, Run), String> {
        match selected(picks, "compare") {
            "machine vs machine" => Ok((
                super::require_run(store, selected(picks, "host-a"))?,
                super::require_run(store, selected(picks, "host-b"))?,
            )),
            "pick two runs" => Ok((
                super::require_run(store, selected(picks, "run-a"))?,
                super::require_run(store, selected(picks, "run-b"))?,
            )),
            _ => {
                let (kind, _) = Self::grouping(picks);
                let host = self.host(picks);
                let find = |side: &str| {
                    let key = selected(picks, &format!("{kind}-{side}"));
                    self.runs
                        .iter()
                        .find(|run| {
                            run.host == host
                                && if kind == "epoch" {
                                    run.epoch() == key
                                } else {
                                    run.os_id() == key
                                }
                        })
                        .cloned()
                        .ok_or_else(|| "selected run is no longer available".to_string())
                };
                Ok((find("a")?, find("b")?))
            }
        }
    }
}
pub fn open(store: &Store, flow: &str) -> Result<(), String> {
    super::require_terminal(if flow.is_empty() {
        "benchmark menu"
    } else {
        flow
    })?;
    let history = History {
        runs: store.list_runs(None, ANY)?,
        hosts: store.known_hosts()?,
    };
    let Some(picks) = cascade("sysinfo bench", |picks| history.expand(flow, picks))? else {
        return Ok(());
    };
    if picks.last().is_some_and(|p| p.kind == "note") {
        return Err(picks.last().unwrap().option.clone());
    }
    let name = if flow.is_empty() {
        selected(&picks, "menu")
    } else {
        flow
    };
    match name {
        "show" => {
            report::render_run(&super::require_run(store, selected(&picks, "run"))?);
            Ok(())
        }
        "compare" => {
            let (a, b) = history.sides(store, &picks)?;
            super::emit_comparison(&a, &b, false)
        }
        "trend" => {
            let runs = store.list_runs(Some(history.host(&picks)), CLEAN)?;
            report::render_trend(&runs, selected(&picks, "metric"));
            Ok(())
        }
        "baseline" => {
            super::set_baseline(store, &super::require_run(store, selected(&picks, "run"))?)
        }
        _ => {
            let matches = super::command()
                .try_get_matches_from(["bench", name])
                .map_err(|e| e.to_string())?;
            super::run(&matches)
        }
    }
}
